sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,535 @@
+  return EMMC_BUSWIDTH_1;
+}
+
+#define MAXBLKSIZE 512 /* Currently only blocks of 512 bytes are supported. */
+#define MAXBLKCNT 255  /* Transfers cannot exceed 255 consecutive blocks. */
+
+static int bcmemmc_get_prop(device_t *cdev, uint32_t id, uint64_t *var) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)cdev->parent->state;
+  resource_t *emmc = state->emmc;
+
+  switch (id) {
+    case EMMC_PROP_RW_BLKCNT:
+      *var = (b_in(emmc, BCMEMMC_BLKSIZECNT) & BSC_BLKCNT) >> BSC_BLKCNT_SHIFT;
+      break;
+    case EMMC_PROP_RW_BLKSIZE:
+      *var = b_in(emmc, BCMEMMC_BLKSIZECNT) & BSC_BLKSIZE;
+      break;
+    case EMMC_PROP_R_MAXBLKSIZE:
+      *var = MAXBLKSIZE;
+      break;
+    case EMMC_PROP_R_MAXBLKCNT:
+      *var = MAXBLKCNT;
+      break;
+    case EMMC_PROP_R_VOLTAGE_SUPPLY:
+      *var = EMMC_VOLTAGE_WINDOW_LOW;
+      break;
+    case EMMC_PROP_RW_RESP_LOW:
+      *var = (uint64_t)b_in(emmc, BCMEMMC_RESP0) |
+             (uint64_t)b_in(emmc, BCMEMMC_RESP1) << 32;
+      break;
+    case EMMC_PROP_RW_RESP_HI:
+      *var = (uint64_t)b_in(emmc, BCMEMMC_RESP2) |
+             (uint64_t)b_in(emmc, BCMEMMC_RESP3) << 32;
+      break;
+    case EMMC_PROP_RW_BUSWIDTH:
+      *var = bcmemmc_get_bus_width(state);
+      break;
+    case EMMC_PROP_RW_RCA:
+      *var = state->rca;
+      break;
+    default:
+      return EINVAL;
+  }
+
+  return 0;
+}
+
+static int bcmemmc_set_prop(device_t *cdev, uint32_t id, uint64_t var) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)cdev->parent->state;
+  resource_t *emmc = state->emmc;
+
+  uint32_t reg = 0;
+  switch (id) {
+    case EMMC_PROP_RW_BLKCNT:
+      if (var >= 65536)
+        return EINVAL;
+      reg = b_in(emmc, BCMEMMC_BLKSIZECNT) & ~BSC_BLKCNT;
+      b_out(emmc, BCMEMMC_BLKSIZECNT, reg | (var << BSC_BLKCNT_SHIFT));
+      break;
+    case EMMC_PROP_RW_BLKSIZE:
+      if (var >= 1024)
+        return EINVAL;
+      reg = b_in(emmc, BCMEMMC_BLKSIZECNT) & ~BSC_BLKSIZE;
+      b_out(emmc, BCMEMMC_BLKSIZECNT, reg | var);
+      break;
+    case EMMC_PROP_RW_RESP_LOW:
+      b_out(emmc, BCMEMMC_RESP0, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP1, (uint32_t)(var >> 32));
+      break;
+    case EMMC_PROP_RW_RESP_HI:
+      b_out(emmc, BCMEMMC_RESP2, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP3, (uint32_t)(var >> 32));
+      break;
+    case EMMC_PROP_RW_CLOCK_FREQ:
+      return bcmemmc_set_clk_freq(cdev->parent, var);
+    case EMMC_PROP_RW_BUSWIDTH:
+      return bcmemmc_set_bus_width(state, var);
+    case EMMC_PROP_RW_RCA:
+      state->rca = var;
+      break;
+    default:
+      return EINVAL;
+  }
+
+  return 0;
+}
+
+static int bcmemmc_send_cmd(device_t *cdev, emmc_cmd_t cmd, uint32_t arg,
+                            emmc_resp_t *resp) {
+  device_t *dev = cdev->parent;
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+  uint32_t code = bcmemmc_encode_cmd(cmd);
+
+  if (cmd.flags & EMMC_F_APP)
```

Comment:
Please put comment here why it is needed. If possible refer reader to given page of documentation.

Reply:
Done. I do, however, believe that anyone who is remotely familiar with e.MMC specs should be able to recognize this pattern.